speechbrain/utils/distributed_metrics.py

```diff
@@ -0,0 +1,263 @@
+"""SpeechBrain utility functions for distributed metrics.
```

Comment:
Hum ... I struggle to see why we actually need all this. Are we using every single function in this file? If so, couldn't we just simplify it given our use case? it reads as a complicated file to do something that should be fairly simple to do? In particulare, why do we need to have something recursive here?

Reply:
This is a valid remark. However, I don't see a way to make it lighter without harming the provided features. The purpose of using `recursively_apply` is to synchronize tensors/objects without having prior knowledge of the input data structure. If we know that we will always have a list of tensors, then it's easy; you just need to call `all_gather_into_tensor`. However, we do not have a standardized way of using metrics in SpeechBrain. Some metrics will expect a list of lists of strings, while some will expect a list of tensors. To cover such a variety, it is required to have a function that can work with arbitrary input data. The purpose of `recursively_apply` is really to first explore the data structure and to apply some function only on some objects. For instance, one purpose is to check if all the nested input data are tensors or not. If so, then you can simply call `all_gather_into_tensor`. Another is to apply the `reduce` function only on tensors. This makes things easier to deal with. Other than that, the remaining function aims to deal with objects or tensors. If you encounter a tensor, you need to use the appropriate function since there can be some mismatch with Gloo vs NCCL.

Ping @pplantinga, I'd like to get your feedback on that! thanks. 

BTW --- I will create a new `distributed_utils.py` file since `recursively_apply` and some other functions should be there because they are used in other context like for the `reduce` fn.

Reply:
Note: we see similar design patterns in other toolkits like ESPNet (https://github.com/espnet/espnet/blob/d561f1eaf1deac2aec3b86b68480de7c084f3f6e/espnet2/torch_utils/recursive_op.py#L9) but I think the way of doing it in speechbrain (hence in HF accelerate) is much better since you can reuse the code in other places without having to copy and paste some boilerplate code.

Reply:
I think @TParcollet  was commenting on `distributed_metrics.py` rather than `recursively_apply`. And while I'm not sure about this specific file, it does seem there may be some unnecessary complexity here. See overview comment for more thoughts.

Reply:
> I think @TParcollet was commenting on `distributed_metrics.py` rather than `recursively_apply`. And while I'm not sure about this specific file, it does seem there may be some unnecessary complexity here. See overview comment for more thoughts.

It was related to this file. I moved `recursively_apply` in a different file but at the beginning `recursively_apply` was in `distributed_metrics.py` hence titouan's comment